grails-app/controllers/org/pih/warehouse/notification/NotificationController.groovy

```diff
@@ -104,63 +90,206 @@ class NotificationController {
+                requisition.dateRequested = dateRequested
+            } else {
+                requisition.requestedDeliveryDate = new Date()
+            }
+        } catch (Exception ex) {
+            log.error "Error in reading dateRequested from order json:${oJson}"
+            requisition.requestedDeliveryDate = new Date()
+        }
+        Boolean lineItemHasError = false
+        JSONArray orderLineItemsArray = oJson.getJSONArray("orderItems")
+        for (int lIndex = 0; lIndex < orderLineItemsArray?.length(); lIndex++) {
+            JSONObject lineItem = orderLineItemsArray.getJSONObject(lIndex)
+            log.info "Reading LineItem:${lineItem} for Order"
             try {
-                if (!productInstance?.validate() || productInstance?.hasErrors()) {
-                    productInstance?.errors?.allErrors?.each {
-                        log.error("ERR:${it}")
+                RequisitionItem requisitionItem = new RequisitionItem()
+                Product product = Product.findByUpc(lineItem.getString("productId"))
+                if (!product) {
+                    JSONObject productMapping = toJson(grailsApplication.config?.openboxes?.integration?.productMapping ?: null)
+                    log.info "ProductMapping:${productMapping}"
+                    if(productMapping) {
+                        String productId = lineItem.getString("productId")
+                        log.info "Reading productId:${productId}"
+                        if(productMapping.has(productId)) {
+                            product = Product.findByProductCode(productMapping.getString(productId))
+                        }
                     }
-                } else {
-                    productInstance.save(flush: true, failOnError: true)
                 }
+                log.info "product:${product} for lineItem index:${lIndex}"
+                Integer quantity = lineItem.getInt("quantity")
+                requisitionItem.product = product
+                requisitionItem.quantity = quantity
+                requisition.addToRequisitionItems(requisitionItem)
             } catch (Exception ex) {
-                log.error "Error in saving PRODUCT:${ex?.message}"
+                log.error "Error:${ex.printStackTrace()} in reading Order Line Item:${lineItem}"
+                lineItemHasError = true
+            }
+        }
+        try {
+            if (!lineItemHasError) {
+                if (requisition.validate() && !requisition.hasErrors()) {
+                    requisition.save(flush: true, failOnError: true)
+                } else {
+                    requisition?.errors?.allErrors?.each {
+                        log.error("ERR:${it}")
+                    }
+                }
+            }
+        } catch (Exception ex) {
+            log.error "Error:${ex.printStackTrace()} in saving ORDER:${ex?.message}"
+        }
+//        }
+    }
+
+    JSONObject toJson(String value) {
```

Comment:
We should be able to get rid of these next two methods now.

Reply:
we need toJson for requestedBy user. Let me know if we need to remove this mapping also.